### PR TITLE
Create a simple generic GraphViz renderer for DiGraph

### PR DESCRIPTION
**Pull request by chick**
There are three basic kinds
- A simple default renderer
- A ranked renderer that places nodes in columns based on depth from sources
- A sub-graph render for graphs that contain a loop
  - Renders just nodes that are part of first loop found
  - Plus the neighbors of the loop
  - Loop edges are shown in red.

### CODE REVIEW

File: src/main/scala/firrtl/graph/RenderDiGraph.scala [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    * val rend = new RenderDiGraph(graph, "alice") {
+    *   override def renderNode(node: Symbol): String = s"\"${symbol.name}\""
+    *   }
+    * }}}
+    */
+  def renderNode(node: T): String = {
+    s""""${node.toString}""""
+  }
+
+  def findOneLoop: Set[T] = {
```
Comment (edwardcwang) [inline src/main/scala/firrtl/graph/RenderDiGraph.scala]: Missing scaladoc?

File: src/main/scala/firrtl/graph/RenderDiGraph.scala [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      newRenderer.toDotWithLoops(loop, getRankedNodes)
+    }
+    else {
+      ""
+    }
+  }
+
+  /**
+    * returns a string that is a graphviz digraph
+    * @return
```
Comment (edwardcwang) [inline src/main/scala/firrtl/graph/RenderDiGraph.scala]: Maybe put the return in the return field?

File: src/main/scala/firrtl/graph/RenderDiGraph.scala [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        s.append(s"""  ${renderNode(parent)} -> ${renderNode(child)};""" + "\n")
+      }
+    }
+    s.append("}\n")
+    s.toString
+  }
+
+  /**
+    * returns a string that is a graphviz digraph, but with loops highlighted
+    * @return
```
Comment (edwardcwang) [inline src/main/scala/firrtl/graph/RenderDiGraph.scala]: Maybe put the return in the return field?

File: src/main/scala/firrtl/graph/RenderDiGraph.scala [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+    paredRankedNodes.foreach { nodesAtRank =>
+      s.append(s"""  { rank=same; ${nodesAtRank.map(renderNode).mkString(" ")} };""" + "\n")
+    }
+
+    s.append("}\n")
+    s.toString
+  }
+
+  def getRankedNodes: mutable.ArrayBuffer[Seq[T]] = {
```
Comment (edwardcwang) [inline src/main/scala/firrtl/graph/RenderDiGraph.scala]: Missing scaladoc

File: src/main/scala/firrtl/graph/RenderDiGraph.scala [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+    walkByRank(diGraph.findSources.toSeq)
+    rankNodes
+  }
+  /**
+    * returns a string that is a graphviz digraph
+    * this method may look better than the toDot format. It tries to align nodes in columns based
+    * on their minimum distance to a source.
+    * Can also be faster and better behaved on large graphs
+    * @return
```
Comment (edwardcwang) [inline src/main/scala/firrtl/graph/RenderDiGraph.scala]: ditto

File: src/main/scala/firrtl/graph/RenderDiGraph.scala (hunk 1 above)
Comment (chick) [reply src/main/scala/firrtl/graph/RenderDiGraph.scala]: Done, moved actual loop finding into DiGraph

File: src/main/scala/firrtl/graph/RenderDiGraph.scala (hunk 2 above)
Comment (chick) [reply src/main/scala/firrtl/graph/RenderDiGraph.scala]: Done

File: src/main/scala/firrtl/graph/RenderDiGraph.scala (hunk 3 above)
Comment (chick) [reply src/main/scala/firrtl/graph/RenderDiGraph.scala]: Done

File: src/main/scala/firrtl/graph/RenderDiGraph.scala (hunk 4 above)
Comment (chick) [reply src/main/scala/firrtl/graph/RenderDiGraph.scala]: Done

File: src/main/scala/firrtl/graph/RenderDiGraph.scala (hunk 5 above)
Comment (chick) [reply src/main/scala/firrtl/graph/RenderDiGraph.scala]: Done

**Comment by albert-magyar** [comment]
@chick do you have an example of the "self path" behavior vs. what you want to see?

**Comment by chick** [comment]
This is what I think is the compelling example, add this to DiGraphTests
```scala
  "DiGraph#path(a,a)" should "(a,b,d,a) or (a,c,d,a)" in {
    cyclicGraph.path("a", "a").length should be (4)
  }
```

**Comment by albert-magyar** [comment]
Finding SCCs is definitely a good way to find loops, but it comes with the caveats that a) it gives SCCs, so it won't actually return a path from one specified node in the SCC back to itself and b) it runs on the whole graph, finding all SCCs at once.

In principle, the API of `path` is pretty loose: it gives you one path. In this case, the implementation is BFS, which is efficient and gives short paths. I'm not really inclined to add an extra restriction (paths must be nontrivial), as it has an undesirable side effect of making the case of `path(v,v)` in an _acyclic_ graph ill-defined and/or inconsistent.

**Comment by chick** [comment]
This is not a big deal to implement by starting at the node in question's children, but why does a node have a path to itself if there is not an edge connecting it to itself

**Comment by albert-magyar** [comment]
My understanding is that paths with zero edges are generally accepted as paths per most definitions. As far as I can tell, trails with one repeated edge (aka simple cycles) are less universally included in the definition of paths.

**Comment by chick** [comment]
@edwardcwang I don't think this PR is going to wait or implement any more functionality to DiGraph, in particular, with the way path is interpreted and computed. I think I have fixed your other comments, when you get the chance can you check, confirm, and approve.

**Pull request merged**